src/kpi/statistics/couplingOfComponents/lib/prFilesUtil.ts

```diff
@@ -8,79 +10,110 @@ import { coupling } from '../model/coupling';
+          registerSubset(combinationMap, key);
+        }
+        countOccurence(combinationMap, key, pullRequest.pullRequest);
       }
-    });
-  });
-
-  deleteOccurences(occs);
-
-  // sorting map entries descending by occurences
-  const res = [...combinationMap.entries()].sort(
-    (a, b) => b[1]['occs'] - a[1]['occs'],
-  );
+    }
+    filterIrrelevantSubsets(combinationMap, minOccurences);
 
-  return res;
+    if (combinationMap.size) {
+      const avgCoupling = calculateAverage(combinationMap);
+      result.data.push({
+        _id: interval._id,
+        avg: avgCoupling,
+        data: [...Object.entries(transformMapToObject(combinationMap))],
+      });
 
-  function increaseCounterAndAddPR(key: string, prId: string) {
-    const currEntry = combinationMap.get(key);
-    currEntry.occs += 1;
-    currEntry.pullrequests.push(prId);
+      couplingTotal += avgCoupling * combinationMap.size;
+      subsetTotal += combinationMap.size;
+    }
   }
+  result.avg = couplingTotal / subsetTotal;
+  return result;
+}
 
-  function addFileCombinationToMap(key: string, pr: string) {
-    combinationMap.set(key, { occs: 1, pullrequests: [pr] });
-  }
+function registerSubset(map: CombinationMap, subset: string) {
+  map.set(subset, { occurences: 0, pullRequests: [] });
+}
 
-  function deleteOccurences(occs: number) {
-    for (let [key, val] of combinationMap.entries()) {
-      if (val.occs < occs) {
-        combinationMap.delete(key);
-      }
+function countOccurence(
+  map: CombinationMap,
+  subset: string,
+  pullRequest: PullRequestDocument,
+) {
+  const entry = map.get(subset);
+  entry.occurences += 1;
+  entry.pullRequests.push(pullRequest._id.toString());
+}
+
+function filterIrrelevantSubsets(map: CombinationMap, minOccurences: number) {
+  for (const [key, entry] of map.entries()) {
+    if (entry.occurences < minOccurences) {
+      map.delete(key);
     }
   }
 }
 
-/**
- * Computes all binomial subsets of @param arr with @param size.
- * I.e. all combinations without order and without repetition.
- * @returns an array with all combination arrays, if input array
- * is greater than size. If same size, it returns the input array,
- * else it returns an empty erray.
- */
-function binomialSubsets(arr: string[], size: number): string[][] {
-  function combinations(part: string[], start: number) {
-    let result = [];
+function calculateAverage(map: CombinationMap) {
+  let sum = 0;
+  for (const [key, entry] of map.entries()) {
+    sum += entry.occurences;
+  }
+  return sum / map.size;
+}
 
-    for (let i = start; i < arr.length; i++) {
-      let p = [...part]; // get a copy of part
-      p.push(arr[i]); // add the iterated element to p
-      if (p.length < size) {
-        // test if recursion can go on
-        result = result.concat(combinations(p, i + 1)); // call combinations again & concat result
-      } else {
-        result.push(p); // push p to result, stop recursion
```

Comment:
I have an issue regarding your calculation of the binomial subsets, because it misses a lot of combinations now.
Since we want to calculate all `k-size subsets` of a `n-size set` with no respect to the order of the elements in the subset, the `binomial coefficient` gives us the [number of the combinations](https://www.arndt-bruenner.de/mathe/scripts/binkoeff1.htm) we need.

Simple example of `subsets(['a', 'b', 'c', 'd'], 3)` yields 
`[ [ 'a', 'b', 'c' ], [ 'a', 'b', 'd' ], [ 'b', 'c', 'd' ] ]` and therefore misses `['a', 'c', 'd']`. 

And if we compute `5 over 3` we got only 6 subsets instead of 10, and so on...

Because I found myself in that situation realizing it is a non trivial problem to solve this, I used that recursive approach I found, which gathers the right subsets reliable (but which is not as performant as an iterative solution, sadly)

Reply:
Good catch. I came up with a new approach, that works better:

The unique subsets of size `3` of the set `S = ["a", "b", "c", "d", "e"]` are, represented as indices, the following:
```
0 1 2
0 1 3
0 1 4
0 2 3
0 2 4
0 3 4
1 2 3
1 2 4
1 3 4
2 3 4
```
To step from a subset to the next, find the first member of the subset at index `j` that has at least `size - j` successors in `S` and set it to this successor. Then change its successors in the subset  to its successors in `|S|`.  Always start from the right-most member.
Going from `0 3 4` to `1 2 3`, we see that the element `4` does not have `3 - 2 = 1` successors in `S` and is skipped.   
Likewise, element `3` does not have `3 - 1 = 2` successors in `S` and is skipped.  
Element `0`, however, has more than `3 - 0 = 3` successors in `S` and is therefore set to `1`.  
We now set elements `3` and `4` to element `1`'s direct successors `2` and `3` and obtain the new subset.

This algorithm stops, once no changes to the subset occurred.

Feel free to review my implementation.

Reply:
This solution looks nice and it seems to work very well!